llm/kv_cache.go

```diff
@@ -0,0 +1,122 @@
+package llm
+
+import (
+	"fmt"
+	"log/slog"
+	"slices"
+
+	"github.com/ollama/ollama/discover"
+	"github.com/ollama/ollama/envconfig"
+)
+
+// Interface for GGML functionality needed by validation
+type GGMLModel interface {
+	KV() KV
+}
+
+// ValidKVCacheTypes contains all supported KV cache types
+var ValidKVCacheTypes = []string{"f32", "f16", "q8_0", "q4_0"}
+
+// modelSupportsFlashAttention checks if the model supports flash attention
+func modelSupportsFlashAttention(ggml GGMLModel) bool {
+	// Check if it's an embedding model - embedding models don't support flash attention
+	if _, ok := ggml.KV()[fmt.Sprintf("%s.pooling_type", ggml.KV().Architecture())]; ok {
+		return false
+	}
+
+	// Check head counts match and are non-zero
+	headCountK := ggml.KV().EmbeddingHeadCountK()
+	headCountV := ggml.KV().EmbeddingHeadCountV()
+	return headCountK != 0 && headCountV != 0 && headCountK == headCountV
+}
+
+// checkFlashAttentionSupport returns whether flash attention should be enabled and a reason if it's not
+func checkFlashAttentionSupport(ggml GGMLModel, gpus discover.GpuInfoList) (bool, string) {
+	flashAttnRequested := envconfig.FlashAttention()
+	if !flashAttnRequested {
+		return false, "not requested"
+	}
+
+	if !gpus.SupportsFlashAttention() {
+		return false, "not supported by GPU"
+	}
+
+	if !modelSupportsFlashAttention(ggml) {
+		return false, "not supported by model"
+	}
+
+	return true, ""
+}
+
+// kVCacheQuantization determines the appropriate KV cache quantization type based on
+// model characteristics, GPU capabilities, and environment settings
+func kVCacheQuantization(ggml GGMLModel, gpus discover.GpuInfoList) string {
```

Comment:
As a last note I think we can rename this to `isValidKvCacheTypeForModel(kvct string, ggml GGMLModel)` to simplify this and decouple it from `envconfig` and anything flash attention related.

Reply:
I've decoupled both envconfig and discover from kv_cache.go, what do you think?